src/shipchain_common/rpc.py

```diff
@@ -45,25 +45,38 @@ def call(self, method, args=None):
 
         try:
             with TimingMetric('engine_rpc.call', tags={'method': method}) as timer:
-                response_json = settings.REQUESTS_SESSION.post(self.url,
-                                                               data=json.dumps(self.payload, cls=DecimalEncoder),
-                                                               timeout=getattr(settings, 'REQUESTS_TIMEOUT', 270)
-                                                               ).json()
+                response = settings.REQUESTS_SESSION.post(self.url, data=json.dumps(self.payload, cls=DecimalEncoder),
+                                                          timeout=getattr(settings, 'REQUESTS_TIMEOUT', 270))
+                response_json = response.json()
+
                 LOG.info('rpc_client(%s) duration: %.3f', method, timer.elapsed)
 
-            if 'error' in response_json:
+            if 'error' in response_json and response.status_code == status.HTTP_200_OK:
+                # It's an error properly handled by engine
                 log_metric('engine_rpc.error', tags={'method': method, 'code': response_json['error']['code'],
                                                      'module': __name__})
                 LOG.error('rpc_client(%s) error: %s', method, response_json['error'])
                 raise RPCError(response_json['error']['message'])
 
+            if status.HTTP_400_BAD_REQUEST <= response.status_code:
+                # It's an unexpected error not handled by engine
+                log_metric('engine_rpc.error', tags={'method': method, 'code': response.status_code,
+                                                     'module': __name__})
+                LOG.error('rpc_client(%s) error: %s', method, response_json)
+                raise RPCError(str(response_json), status_code=response.status_code)
+
             response_json = response_json['result']
 
         except requests.exceptions.ConnectionError:
             # Don't return the true ConnectionError as it can contain internal URLs
             log_metric('engine_rpc.error', tags={'method': method, 'code': 'ConnectionError', 'module': __name__})
-            raise RPCError("Service temporarily unavailable, try again later", status.HTTP_503_SERVICE_UNAVAILABLE,
-                           'service_unavailable')
+            raise RPCError("Service temporarily unavailable, try again later",
+                           status_code=status.HTTP_503_SERVICE_UNAVAILABLE, code='service_unavailable')
+
+        except RPCError as rpc_exc:
+            # Since we are in a try ... except loop, the above raised
+            # rpc errors need to be re-raised here in order to be properly catch
```

Comment:
Do you have a source for this? I don't think this is true. Raising an exception in an `except` block is valid, it shouldn't need to be rethrown. 
https://docs.python.org/3.6/tutorial/errors.html#raising-exceptions

Also, this `except` block won't get triggered by the `raise` on ln73. Exceptions thrown within `except` blocks won't get handled by other adjacent `except` blocks. The `except` blocks only handle exceptions for the contents of the `try` block

Reply:
You are right nothing forbid to just raise those exceptions in the `try` block. 
However, in this particular case, the `RPCError` thrown in the `try` block will get caught by the generic `Exception` in line 81 and we will loose some information like the `status_code` of the that exception. That is why I opted for re-raising them here.

Reply:
I see, I didn't realize we were throwing RPCError in the try block with different status_codes. In that case, I think this block is best simplified as
```
except RPCError:
    raise
```
which will preserve the call stack of the original exception and re-raise, instead of letting the fallback error handler handle it. Please either delete those comments or reword them so that they accurately explain the need for this.